src/main/java/com/microsoft/azure/util/AzureMsiCredentials.java

```diff
@@ -2,24 +2,87 @@
 
 import com.cloudbees.plugins.credentials.CredentialsScope;
 import com.cloudbees.plugins.credentials.impl.BaseStandardCredentials;
+import com.microsoft.azure.AzureEnvironment;
 import hudson.Extension;
+import hudson.util.ListBoxModel;
+import org.apache.commons.lang.StringUtils;
 import org.kohsuke.stapler.DataBoundConstructor;
 
+import java.util.Map;
+
 public class AzureMsiCredentials extends BaseStandardCredentials {
 
     public static final int DEFAULT_MSI_PORT = 50342;
+    private static final long serialVersionUID = 6743945878507124459L;
+
     private final int msiPort;
+    private String azureEnvName;
+    private transient AzureEnvironment azureEnvironment;
 
     @DataBoundConstructor
-    public AzureMsiCredentials(CredentialsScope scope, String id, String description, int msiPort) {
+    public AzureMsiCredentials(CredentialsScope scope, String id, String description, int msiPort,
+                               String azureEnvName) {
         super(scope, id, description);
         this.msiPort = msiPort;
+        this.azureEnvName = azureEnvName;
+
+        resolveAzureEnv();
+    }
+
+    private void resolveAzureEnv() {
+        if (AzureCredentials.Constants.ENV_AZURE.equalsIgnoreCase(azureEnvName)) {
+            azureEnvironment = AzureEnvironment.AZURE;
+        } else if (AzureCredentials.Constants.ENV_AZURE_CHINA.equalsIgnoreCase(azureEnvName)) {
+            azureEnvironment = AzureEnvironment.AZURE_CHINA;
+        } else if (AzureCredentials.Constants.ENV_AZURE_GERMANY.equalsIgnoreCase(azureEnvName)) {
+            azureEnvironment = AzureEnvironment.AZURE_GERMANY;
+        } else if (AzureCredentials.Constants.ENV_AZURE_US_GOVERNMENT.equalsIgnoreCase(azureEnvName)) {
+            azureEnvironment = AzureEnvironment.AZURE_US_GOVERNMENT;
+        } else {
+            azureEnvironment = AzureEnvironment.AZURE;
+        }
+    }

```

Comment:
We have similar logic in `AzureCredentials.ServicePrincipal#getAzureEnvironment()`. Could you extract the logic to a helper method?